hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -0,0 +1,204 @@
+        return {'gs'}
+
+    @staticmethod
+    def _get_bucket_name(url: str) -> Tuple[str, str]:
+        parsed = urllib.parse.urlparse(url)
+        if parsed.scheme != 'gs':
+            raise ValueError(f"invalid scheme, expected gs: {parsed.scheme}")
+
+        name = parsed.path
+        if name:
+            assert name[0] == '/'
+            name = name[1:]
+
+        return (parsed.netloc, name)
+
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        if not all(c in 'rwxabt+' for c in mode):
+            raise ValueError(f"invalid mode: {mode}")
+        if 't' in mode and 'b' in mode:
+            raise ValueError(f"can't have text and binary mode at once: {mode}")
+        if 'b' not in mode:
+            raise ValueError(f"text mode not supported: {mode}")
+        if 'x' in mode:
+            raise ValueError(f"exclusive creation not supported: {mode}")
+        if 'a' in mode:
+            raise ValueError(f"append mode not supported: {mode}")
+        if '+' in mode:
+            raise ValueError(f"updating not supported: {mode}")
+        if ('r' in mode) + ('w' in mode) != 1:
+            raise ValueError(f"must have exactly one of read/write mode: {mode}")
+
+        bucket, name = self._get_bucket_name(url)
+
+        if 'r' in mode:
+            return await self._storage_client.get_object(bucket, name)
+
+        assert 'w' in mode
+        return await self._storage_client.insert_object(bucket, name)
+
+    async def mkdir(self, url: str) -> None:
+        pass
+
+    async def isfile(self, url: str) -> bool:
+        try:
+            bucket, name = self._get_bucket_name(url)
+            await self._storage_client.get_object_metadata(bucket, name)
+            return True
+        except aiohttp.ClientResponseError as e:
+            if e.status == 404:
+                return False
+            raise
+
+    async def isdir(self, url: str) -> bool:
+        bucket, name = self._get_bucket_name(url)
+        if not name.endswith('/'):
+            name = f'{name}/'
+        params = {
+            'delimiter': '/',
+            'includeTrailingDelimiter': 'true',
+            'maxResults': 1,
+            'prefix': name
+        }
+        data = await self._storage_client._list_objects(bucket, params=params)
+        assert 'prefixes' not in data or data['prefixes'] is None
+        items = data.get('items')
+        return items is not None and len(items) > 0
+
+    async def remove(self, url: str) -> None:
+        bucket, name = self._get_bucket_name(url)
+        await self._storage_client.delete_object(bucket, name)
+
+    async def rmtree(self, url: str) -> None:
+        bucket, name = self._get_bucket_name(url)
+        if not name.endswith('/'):
+            name = f'{name}/'
+        params = {
+            'prefix': name
+        }
+        done = False
+        while not done:
+            done = True
+            data = await self._storage_client._list_objects(bucket, params=params)
```

Comment:
Is there a page that list_objects returns? I would have thought this should be an async generator or have some sort of params option that lists the maximum number of results. Otherwise, a while loop here doesn't make sense.

Reply:
Yes.  I haven't implemented that yet, I was planning to do it as part of adding a high level list objects interface.  This works without the paging by deleting the objects on the first page (the API defaults to 1000 items), and then listing again after deleting those.  Since we deleted, when we list again, we'll get a new page of objects without using the next page token.